lib/p2psocket.js

```diff
@@ -77,93 +101,200 @@ P2PSocket.prototype.parseInv = function(payload) {
       type: payload.readUInt32LE(offset),
-      hash: payload.slice(offset + 4, offset + 4 + 32)
+      hash: inPlaceReverse(payload.slice(offset + 4, offset + 4 + 32)).toString('hex')
     });
   }
 
   return content;
 }
 
+P2PSocket.prototype.parseBlock = function(payload) {
+
+  var header = payload.slice(0,80);
+
+  var sha256_pass1 = crypto.createHash('sha256').update(header).digest();
+  var blockhash = inPlaceReverse(crypto.createHash('sha256').update(sha256_pass1).digest()).toString('hex');
+
+  var content = {
+    hash:              blockhash,
+    confirmations:     undefined,
+    size:              payload.length,
+    height:            undefined,
+    version:           payload.readUInt32LE(0),
+    merkleroot:        inPlaceReverse(payload.slice(36,68)).toString('hex'),
+    tx:                [],
+    time:              payload.readUInt32LE(68),
+    nonce:             payload.readUInt32LE(76),
+    bits:              inPlaceReverse(payload.slice(72,76)).toString('hex'),
+    difficulty:        undefined,
+    chainwork:         undefined,
+    previousblockhash: inPlaceReverse(payload.slice(4,36)).toString('hex'),
+  };
+
+  var txc = payload.readVarInt(80);
+
+  var i, base = 80+txc.size;
+
+  for ( i=0; i<txc.value; i++ ) {
+    var tx = this.parseTx(new P2PBuffer(payload.slice(base)));
+    content.tx.push(tx);
+    base +=  tx.hex.length / 2;
+  }
+
+
+  /*
+   * Get the bip-0034 height if the block version is greater than 1
+   * https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki
+   */
+  if ( content.version > 1 ) {
+    var coinbase = content.tx[0];
+    content.height = 0;
+
+    var script = new Buffer(coinbase.vin[0].coinbase,'hex');
+    var s = script.readUInt8(0);
+    for (i=0; i<s && i<3; i++) {
+      content.height += script.readUInt8(i+1) * ( i > 0 ? i*256 : 1);
+    }
+  }
+
+
+  return content;
+}
+
 P2PSocket.prototype.parseTx = function(payload) {
+
   var content = {
     version: payload.readUInt32LE(0),
+    hex:  null,
+    txid: null,
+    lock_time: null,
     vin:  [],
     vout: [],
   };
-  
+
   var i = 4;
-  
+
   var cin = payload.readVarInt(i);
   i += cin.size;
-  
+
   var n;
   for ( n=0; n < cin.value; n++ ) {
-  
+
     var tx = {
-      previous_output: {
-        txid: payload.slice(i, i+32),
-        vout: payload.readUInt32LE(i+32),
-      },
+      txid: inPlaceReverse(payload.slice(i, i+32),true).toString('hex'),
+      vout: payload.readUInt32LE(i+32)
     };
-    
+
     i += 36;
-    
+
     var sl = payload.readVarInt(i);
     i += sl.size;
-    
-    tx.scriptSig = payload.slice(i, i+sl.value);
+
+    var scriptSig = payload.slice(i, i+sl.value);
+    tx.scriptSig = {
+      hex: scriptSig.toString('hex'),
+      asm: scriptToAsm(scriptSig),
+    }
+
     i += sl.value;
-    
+
     tx.sequence = payload.readUInt32LE(i);
     i += 4;
-    
-    content.vin.push(tx);      
+
+    if ( tx.vout === tx.sequence && tx.txid.match(/^0+$/) ) {
```

Comment:
Maybe we could replace it with a safer check, because I don't think there's any requirement that `tx.vout === tx.sequence` in a coinbase, and if a coinbase transaction on a live net didn't meet that, the block parser wouldn't get it. Maybe replace it with verifying that the `tx.vout` is `0xffffffff`? That is a requirement of coinbase txs, and is a faster check.